### PR TITLE
GEODE-6267: Logged out subject

### PR DESCRIPTION
**Pull request by boglesby**
The changes to CacheClientProxy  closeTransientFields and closeOtherTransientFields are to address the bug I found while debugging this leak.

The main changes to address the leak are to invoke removeSubject in ClientUserAuths. cleanup and logout in CacheClientProxy.closeOtherTransientFields.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java [hunk 1]
```diff
@@ -963,6 +968,10 @@ private void closeTransientFields() {
       return;
     }
 
+    closeOtherTransientFields();
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java]: I wonder what block 965-967 would do. Can we always close all these transient fields no matter what?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  private static Subject proxySubject;
+
+  private static List<ClientUserAuths> authorizations;
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    Properties locatorProps = new Properties();
+    locatorProps.setProperty(SECURITY_MANAGER, SimpleSecurityManager.class.getCanonicalName());
+    locator = cluster.startLocatorVM(0, locatorProps);
```
Comment (jinmeiliao) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java]: you can also do: `locator = cluster.startLocatorVM(0, l->l.withSecurityManager(SimpleSecurityManager.class));`

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    Properties serverProps = new Properties();
+    serverProps.setProperty("security-username", "cluster");
+    serverProps.setProperty("security-password", "cluster");
+    server = cluster.startServerVM(1, serverProps, locator.getPort());
```
Comment (jinmeiliao) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java]: you can also replace these lines with:
```
server = cluster.startServerVM(1, s->s.withCredential("cluster", "cluster").withConnectionToLocator(locator.getPort()).withRegion(RegionShortcut.PARTITION, "region"));
```

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java [hunk 4]
```diff
@@ -1934,6 +1948,10 @@ public Version getVersion() {
     return this.clientVersion;
   }
 
+  protected Subject getSubject() {
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java]: probably mark this as @VisibleForTesting

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java [hunk 5]
```diff
@@ -78,6 +80,10 @@ public UserAuthAttributes getUserAuthAttributes(long userId) {
     return uniqueIdVsUserAuth.get(userId);
   }
 
+  protected Collection<Subject> getSubjects() {
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java]: probably mark this as @VisibleForTesting

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 6]
```diff
@@ -1102,6 +1102,10 @@ public void removeUserAuth(Message message, boolean keepAlive) {
     }
   }
 
+  protected ClientUserAuths getClientUserAuths() {
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: probably mark this as @VisibleForTesting

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java (hunk 1 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java]: Yes, the transient fields should be closed (or cleaned up) when the CacheClientProxy closes.

They are mostly related to the socket. Transient fields are:

- _socket
- _commBuffer - ByteBuffer on the socket
- releaseResourcesForAddress - removes the asyncCloseExecutors for the proxy's _remoteHostAddress
- clearClientInterestList and closeNonDurableCqs - remove interests and CQs

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java (hunk 2 above)
Comment (boglesby) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java]: Thanks for the suggestion. Fixed.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java (hunk 3 above)
Comment (boglesby) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/ClientServerConnectDisconnectDistributedTest.java]: Thanks for the suggestion. Fixed.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java (hunk 4 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java]: Thanks for the suggestion. Fixed.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java (hunk 5 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java]: Thanks for the suggestion. Fixed.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java (hunk 6 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: Thanks for the suggestion. Fixed.

**Pull request merged**